notes/radicle-p2p-hypercore-overview.md

```diff
@@ -0,0 +1,126 @@
+
+*Single-writer chains* are Radicle chains that only the owner of a specific
+private key can write to.
+
+We use hypercore feeds to represent single-writer chains. Each entry in the feed
+is a Radicle expression. Single-writer chains are identified by the discovery
+key of their feed.
+
+To read a Radicle chain a user needs the discovery key of a feed. With the
+discovery key the user uses hyperdiscovery to find someone who replicates the
+feed. (This may be the owner.) The user then reads all entries in the feed with
+hypercore and evaluates them with radicle.
+
+To write an expression to a Radicle chain a user publishes a new entry to the
+`hypercore` feed.
+
+
+Multi-writer Chains
+-------------------
+
+*Multi-writer chains* are Radicle chains that the owners of multiple distinct
+private keys can write to.
+
+We implement a multi-writer chain by using set of feeds owned by different
+users. Every entry in these feeds is a Radicle expression and a link to another
+entry in any of feeds. Every entry in a feed now determines a linked list of
+Radicle expression (assuming the feed entries and links are well formed and
+valid) and thus a Radicle chain.
+
+To derive a Radicle chain from a set of feeds we need an algorithm that takes
+the heads of all the feeds of a multi-writer chain and outputs either a chain of
+Radicle expressions or signals that the chain is invalid.
+
+The simplest such algorithm is the following: We accept only the situation where
+there is a feed with head entry `e` so that the head entries of all other feeds
+are ancestors of `e`. In that case `e` is considered the head of the chain.
+Otherwise the chain is invalid.
+
+
+**TODO** Properties of this scheme
+
+**TODO** Addressing entries
+
+
+Offline First and Replication
+-----------------------------
+
+**TODO** flesh this out
+
+Hypercore stores every feed a user is interested locally.
+
+Hypercore allows one user to replicate the feed of another user. Updates are
+propagated automatically.
+
+Users can fetch a feed from a user replicating it if the owner is not online.
+
+Implementation
+--------------
+
+`hypercore` and `discovery-swarm` are only available as JavaScript libraries. To
+build Radicle P2P on top of hypercore we would need to implement the core
+functionality of a Radicle P2P node in JavaScript. (Typescript and Purescript
+are alternative languages since they can use JavaScript transparently.) This
+node would expose an API that would allow the Radicle REPL and other tools to
+submit and get expressions.
+
+The node will be accompanied by a control CLI that allows users to create
+chains, replicate chains, etc.
+
+Security
+--------
+
+* Feeds can be censored. If the owner goes offline and all users that replicate
+  the feed conspire to drop part of the head of the feed. If one user chooses to
+  replicate the actual head of the feed then everyone else will see that head,
+  too.
+* Feeds are unforgeable. Anybody with the public key of a feed can verify
```

Comment:
Can the owner forge the feed? E.g. can they change the first entry.

Reply:
If the owner does anything but append an entry to the feed and a version of the previous state of the feed is still being replicated the feed is treated as corrupted.

Feeds provide the same security guarantees as hash linked lists. Changing any entry in the feed will change the overall hash of all following entries and consequently the hash of the current state of the feed.

Reply:
> the feed is treated as corrupted

There isn't really a way to recover from that, is it?